### PR TITLE
Alert acknowledgement

### PR DESCRIPTION
**Pull request by ryanclark**
This takes the work from @fspmarshall here - https://github.com/gravitational/teleport/commit/36da19c1581172bd43c2e11bb6f0eddf35e37a99

This implements the following CLI interface

`tctl alerts ls -v` now displays the ID that's needed for ack'ing an alert

`tctl alerts ack --reason <reason> --ttl <duration> <alert id>`

`tctl alerts ack ls`

`tctl alerts ack --clear <alert id>`

image

### CODE REVIEW

**Comment by zmb3** [comment]
> as the ID and reason are required, i think this should be
> tctl alert ack --ttl=7d upgrade-suggestion "will upgrade next week"

I like positional arguments when there's only one and it's hard to get wrong. But in this example won't it be confusing if people mistakenly swap `upgrade-suggestion` and `will upgrade next week`?

> Is TTL required here? What should the default be if not?

I would expect the default means no expiry, and the ack is valid for as long as the alert exists.

> this doesn't align with our current convention so i think this should be `tctl alert ack ls`

Agreed

> i think we can just make --clear a flag on tctl alert ack so it'll become

That seems fine too.

**Comment by fspmarshall** [comment]
> `tctl alert ack --ttl=7d upgrade-suggestion "will upgrade next week"`

I'd definitely keep `--reason` as a flag.  IMO required flags are fine, and improve readability.  Especially for arguments that may include spaces.

I could go either way for `--id`, but my thinking in making it a required flag in the original proposal was for two reasons:

1. I find commands with many arguments more readable if all of them are labeled.

2. I like the idea of leaving our options open for other kind of acks in the future that don't use ID (e.g. acknowledging all alerts with some particular label).


> Is TTL required here? What should the default be if not?

If I were to pick a default, I'd go with something in the `36h-72h` range.  I think very long suppression should require an explicit ttl.  If you're acknowledging the error to get rid of the annoyance while you fix the issue, then a day or two and change seems good.  If you're acknowledging the error either without intent to fix, or with intent to fix at some later date, then explicit ttl seems better.

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 1]
```diff
@@ -2027,6 +2045,15 @@ service AuthService {
   // UpsertClusterAlert creates a cluster alert.
   rpc UpsertClusterAlert(UpsertClusterAlertRequest) returns (google.protobuf.Empty);
 
+  // AcknowledgeAlert marks a cluster alert as acknowledged.
+  rpc AcknowledgeAlert(types.AlertAcknowledgement) returns (google.protobuf.Empty);
```
Comment (strideynet) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2049]: Nit: Part of me thinks that we could just call this `CreateAlertAck` for consistency with the naming of other CRUD methods and the other AlertAck messages..

**Review by strideynet — APPROVED** [review]
Overall good, just a few thoughts on this.

File: lib/services/local/status.go [hunk 2]
```diff
@@ -146,11 +147,70 @@ func (s *StatusService) DeleteClusterAlert(ctx context.Context, alertID string)
...
 
-const clusterAlertPrefix = "cluster-alerts"
+// AcknowledgeAlert marks a cluster alert as acknowledged.
+func (s *StatusService) AcknowledgeAlert(ctx context.Context, ack types.AlertAcknowledgement) error {
+	if err := ack.Check(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if ack.Expires.IsZero() {
+		ack.Expires = s.Clock().Now().UTC().Add(time.Hour * 24)
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply lib/services/local/status.go]: We discussed this on Slack and as there is a default value set by the tctl CLI, this check has been removed and an error will be thrown if an expiry time isn't set

**Review by strideynet — APPROVED** [review]
Thanks for sorting out the jsontags :')

**Review by avatus — APPROVED** [review]

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**